src/emitter.ts

```diff
@@ -0,0 +1,1126 @@
+    function paramsToString(ps: Browser.Param[]) {
+        function paramToString(p: Browser.Param) {
+            const isOptional = !p.variadic && p.optional;
+            const pType = isOptional ? convertDomTypeToTsType(p) : convertDomTypeToNullableTsType(p);
+            return (p.variadic ? "..." : "") +
+                adjustParamName(p.name) +
+                (isOptional ? "?: " : ": ") +
+                pType +
+                (p.variadic ? "[]" : "");
+        }
+        return ps.map(paramToString).join(", ");
+    }
+
+    function emitCallBackInterface(i: Browser.Interface) {
+        if (i.name === "EventListener") {
+            printer.printLine(`interface ${i.name} {`);
+            printer.printWithAddedIndent("(evt: Event): void;");
+            printer.printLine("}");
+        }
+        else {
+            const methods = mapToArray(i.methods.method);
+            const m = methods[0];
+            const overload = m.signature[0];
+            const paramsString = overload.param ? paramsToString(overload.param) : "";
+            const returnType = overload.type ? convertDomTypeToTsType(overload) : "void";
+            printer.printLine(`type ${i.name} = ((${paramsString}) => ${returnType}) | { ${m.name}(${paramsString}): ${returnType}; };`);
+        }
+        printer.printLine("");
+    }
+
+    function emitCallBackFunction(cb: Browser.CallbackFunction) {
+        printer.printLine(`interface ${cb.name} {`);
+        printer.increaseIndent();
+        emitSignatures(cb, "", "", s => printer.printLine(s));
+        printer.decreaseIndent();
+        printer.printLine("}");
+        printer.printLine("");
+    }
+
+    function emitCallBackFunctions() {
+        getElements(webidl["callback-functions"], "callback-function")
+            .sort(compareName)
+            .forEach(emitCallBackFunction);
+    }
+
+    function emitEnum(e: Browser.Enum) {
+        printer.printLine(`type ${e.name} = ${e.value.map(v => `"${v}"`).join(" | ")};`);
+    }
+
+    function emitEnums() {
+        getElements(webidl.enums, "enum")
+            .sort(compareName)
+            .forEach(emitEnum);
+    }
+
+    function emitEventHandlerThis(prefix: string, i: Browser.Interface) {
+        if (prefix === "") {
+            return `this: ${i.name}, `;
+        }
+        else {
+            return pollutor ? `this: ${pollutor.name}, ` : "";
+        }
+    }
+
+    // A covariant  EventHandler is one that is defined in a parent interface as then redefined in current interface with a more specific argument types
+    // These patterns are unsafe, and flagged as error under --strictFunctionTypes.
+    // Here we know the property is already defined on the interface, we elide its declaration if the parent has the same handler defined
+    function isCovariantEventHandler(i: Browser.Interface, p: Browser.Property) {
+        return isEventHandler(p) &&
+            iNameToEhParents[i.name] && iNameToEhParents[i.name].length > 0 &&
```

Comment:
Nit: could use local variables for `iNameToEhParents[i.name]` and `iNameToEhList[i.name]`. Also avoid shadowing `i` so close to the first definition.